### PR TITLE
HPOS: fix handling order meta containing nonexistent class when sync is on

### LINKED ISSUE #38304: Uncaught error when calling wc_get_orders (no other plugins active)
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

I encountered this after setting up a test site on which I deactivated all plugins except WooCommerce. In summary: having an inactive (or obsolete+deleted) plugin that had stored an object in metadata can result in WC triggering an uncaught PHP fatal error when calling `wc_get_orders()`.

N.B. HPOS is active (I don't know whether this is related).

```
% php -a
php > echo PHP_VERSION; // By the way, 
8.2.5
php > require 'wp-load.php'; // Lots of PHP deprecation notices for dynamic properties in WC, but this issue is about something else. 
php > echo WOOCOMMERCE_VERSION;
7.7.0
php > $order_results = wc_get_orders(array('limit' => 300, 'offset' => 0, 'order' => 'DESC',
'order_by' => 'ID', 'return' => 'objects', 'meta_key' => '_openinghours_time', 'meta_value' => '', 'meta_compare' => '!='));

Warning: Uncaught Error: The script tried to modify a property on an incomplete object. Please ensure that the class definition "geoiprecord" of the object you are trying to operate on was loaded _before_ unserialize() gets called or provide an autoloader to load the class definition in wp-includes/formatting.php:5035
Stack trace:
#0 wp-includes/formatting.php(5030): map_deep(Object(__PHP_Incomplete_Class), 'stripslashes_fr...')
#1 wp-includes/formatting.php(5030): map_deep(Array, 'stripslashes_fr...')
#2 wp-includes/formatting.php(2810): map_deep(Array, 'stripslashes_fr...')
#3 wp-includes/formatting.php(5688): stripslashes_deep(Array)
#4 wp-includes/meta.php(53): wp_unslash(Array)
#5 wp-includes/post.php(2446): add_metadata('post', 70, 'vat_compliance_...', Array, false)
#6 wp-content/plugins/woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php(623): add_post_meta(70, 'vat_compliance_...', Array, false)
#7 wp-content/plugins/woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php(560): Abstract_WC_Order_Data_Store_CPT->update_order_meta_from_object(Object(Automattic\WooCommerce\Admin\Overrides\Order))
#8 wp-content/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php(549): Abstract_WC_Order_Data_Store_CPT->update_order_from_object(Object(Automattic\WooCommerce\Admin\Overrides\Order))
#9 wp-content/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php(2156): Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore->backfill_post_record(Object(Automattic\WooCommerce\Admin\Overrides\Order))
#10 wp-content/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php(2489): Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore->maybe_backfill_post_record(Object(Automattic\WooCommerce\Admin\Overrides\Order))
#11 wp-content/plugins/woocommerce/includes/class-wc-data-store.php(224): Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore->update_meta(Object(Automattic\WooCommerce\Admin\Overrides\Order), Object(WC_Meta_Data))
#12 wp-content/plugins/woocommerce/includes/abstracts/abstract-wc-data.php(681): WC_Data_Store->__call('update_meta', Array)
#13 wp-content/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php(1232): WC_Data->save_meta_data()
#14 wp-content/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php(1317): Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore->migrate_meta_data_from_post_order(Object(Automattic\WooCommerce\Admin\Overrides\Order), Object(WC_Order))
#15 wp-content/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php(1120): Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore->migrate_post_record(Object(Automattic\WooCommerce\Admin\Overrides\Order), Object(WC_Order))
#16 wp-content/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php(1026): Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore->maybe_sync_order(Object(Automattic\WooCommerce\Admin\Overrides\Order), Object(WC_Order))
#17 wp-content/plugins/woocommerce/includes/class-wc-data-store.php(171): Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore->read_multiple(Array)
#18 wp-content/plugins/woocommerce/includes/class-wc-order-factory.php(123): WC_Data_Store->read_multiple(Array)
#19 wp-content/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php(2316): WC_Order_Factory::get_orders(Array)
#20 wp-content/plugins/woocommerce/includes/class-wc-data-store.php(224): Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore->query(Array)
#21 wp-content/plugins/woocommerce/includes/class-wc-order-query.php(87): WC_Data_Store->__call('query', Array)
#22 wp-content/plugins/woocommerce/includes/wc-order-functions.php(69): WC_Order_Query->get_orders()
#23 php shell code(1): wc_get_orders(Array)
#24 {main}
  thrown in wp-includes/formatting.php on line 5035
```

Looks similar in concept to (fixed) https://github.com/woocommerce/woocommerce/issues/34913 - WC is running metadata through formatting functions, and those formatting functions can end up loading objects, resulting in uncaught errors. In this specific case the thing it was trying to operate on was:

```
O:11:"geoiprecord":14:{s:12:"country_code";s:2:"BE";s:13:"country_code3";s:3:"BEL";s:12:"country_name";s:7:"Belgium";s:6:"region";s:3:"BRU";s:4:"city";s:8:"Brussels";s:11:"postal_code";s:4:"1000";s:8:"latitude";d:50.833300000000001;s:9:"longitude";d:4.3333000000000004;s:9:"area_code";N;s:8:"dma_code";N;s:10:"metro_code";N;s:14:"continent_code";s:2:"EU";s:11:"region_name";s:16:"Brussels Capital";s:8:"timezone";s:15:"Europe/Brussels";}}
```

### Expected behavior

Handled nicely, no crash. (e.g. catch the error).

### Actual behavior

PHP uncaught error, crash.

### Steps to reproduce

See above.

### WordPress Environment

WP 6.2, other versions as above.

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by lsinger**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Assume a plugin adds metadata to an order, and that metadata is a serialized object that uses a class that the plugin defines. Now assume the plugin is subsequently deactivated and the class definition necessary for deserializing the object in the order's metadata isn't available anymore. 

This is not an issue as long as the metadata isn't touched, but with HPOS *and* sync turned on, there are situations where the metadata is deserialized, changed (by WordPress Core’s `map_deep`), and the serialized again. Under PHP less than 8 this was not an issue, but under PHP 8 throws an error. 

There is [an open ticket](https://core.trac.wordpress.org/ticket/55257) in WordPress Core for this issue but it’s not clear when, how, and if the issue will be addressed. So to ship a fix for WooCommerce users experiencing this issue, this PR catches the error and tries to imitate the metadata actions using `$wpdb` directly. 

Detecting this specific error can, as far as I can tell, only be done by matching the kind of error message, which I find a bit sub-optimal. I'd love to hear ideas for alternative ways of detection. 

Closes #38304.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. ensure CI passes
2. check out the branch locally
3. run the new unit test under PHP 8.0 like so: `nvm use && wp-env stop && WP_ENV_PHP_VERSION=8.0 wp-env start && pnpm test:php:env --filter=test_unserialize_meta_with_nonexistent_class` -- the test should pass
4.  undo the actual fix by walking back a few commits like so: `git checkout 19b2ddf`
5. run the new unit test again: `pnpm test:php:env --filter=test_unserialize_meta_with_nonexistent_class` -- the test should now fail

#### Instructions for Manual Testing

1. Ensure to test this with WooCommerce running on *PHP 8.0 or later*. 
2. Under WooCommerce / Settings / Advanced / Features, ensure *both* HPOS is turned on *and* sync is also turned on. 
3. Install the [Code Snippets](https://wordpress.org/plugins/code-snippets/) plugin.
4. Create an order. 
5. Find the order in wp-admin and note down it's ID. 
6. Add the following snippet, replacing `__ID__` with the ID of the order you just created. Make sure to select the "execute only once" option.

```php
$incomplete_meta_object_test = function() {
	$order = wc_get_order( __ID__ );
	global $wpdb;
	$order_id = $order->get_id();
	$meta_key = 'test_unserialize_meta_with_nonexistent_class';
	$meta_value = 'O:11:"geoiprecord":14:{s:12:"country_code";s:2:"BE";s:13:"country_code3";s:3:"BEL";s:12:"country_name";s:7:"Belgium";s:6:"region";s:3:"BRU";s:4:"city";s:8:"Brussels";s:11:"postal_code";s:4:"1000";s:8:"latitude";d:50.833300000000001;s:9:"longitude";d:4.3333000000000004;s:9:"area_code";N;s:8:"dma_code";N;s:10:"metro_code";N;s:14:"continent_code";s:2:"EU";s:11:"region_name";s:16:"Brussels Capital";s:8:"timezone";s:15:"Europe/Brussels";}}';
	$order_meta_table = Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore::get_meta_table_name();
	$wpdb->query( "INSERT INTO {$order_meta_table} (order_id, meta_key, meta_value) VALUES ({$order_id}, '{$meta_key}', '{$meta_value}')" );
	$order->set_date_modified( time() );
	$order->save();
	$order_results = wc_get_orders(array('limit' => 300,'offset' => 0,'order' => 'DESC','order_by' => 'ID','return' => 'objects','meta_key' => $meta_key,'meta_value' => '','meta_compare' => '!='));
	$meta = $order_results[0]->get_meta( $meta_key );
	$logger = wc_get_logger();
	$logger->error( 'results: ' . var_export( $meta, true ) );
};
add_action( 'woocommerce_after_register_post_type', $incomplete_meta_object_test );
```

7. Save the snippet, and click "Execute Once". 
8. Reload the page. You should not get an error. 
9. Possibly optional (?) -- to ensure the above instructions are correct, retry the same using the latest official WooCommerce. An error should be thrown in the last step after reloading: "Fatal error: Uncaught Error: The script tried to modify a property on an incomplete object. [...]" Another reload should make the error go away. When you look at the error logs (`http://localhost:8888/wp-admin/admin.php?page=wc-status&tab=logs`), the error should be in there as well. 

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

**Comment by Konamiman** [comment]
I get an error when trying to run all the tests of the file (so without the `--filter`):

```
OrdersTableDataStoreTests::test_order_updated_webhook_delivered_once
Error: Class "WC_Tests_Webhook_Functions" not found
```

**Comment by lsinger** [comment]
> I get an error when trying to run all the tests of the file (so without the `--filter`):
> 
> ```
> OrdersTableDataStoreTests::test_order_updated_webhook_delivered_once
> Error: Class "WC_Tests_Webhook_Functions" not found
> ```

@Konamiman Unfortunately I cannot reproduce this locally. Checking out this branch, running `pnpm install`, and then `pnpm test:php:env` gives me (after 411s): 

```
OK, but incomplete, skipped, or risky tests!
Tests: 3971, Assertions: 14354, Skipped: 9.
✔ Ran `vendor/bin/phpunit -c phpunit.xml --verbose` in 'tests-cli'. (in 411s 82ms)
```

Did you run `pnpm install` and possibly `pnpm build` after checking out the branch? I've had issues about files not being found before and that fixed it. My errors looked very different though. 

Note that all the CI checks pass and the class that is not being found in your setup doesn't seem to have anything to do with the changes in this PR -- seems like this could be a local issue?

**Review by vedanshujain — COMMENTED** [review]
Looks good to me, but we need to modify testing instructions so that folks are able to test it without unit test setup. Perhaps we can use a small dummy plugin that adds a class and we can then enable/disable that plugin during testing.

**Comment by Konamiman** [comment]
@lsinger I'm running the test as follows: `vendor/bin/phpunit tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php` (I always run the unit test locally like that). Shouldn't this be equivalent to what you posted in the testing instructions?

**Comment by lsinger** [comment]
> Looks good to me, but we need to modify testing instructions so that folks are able to test it without unit test setup. Perhaps we can use a small dummy plugin that adds a class and we can then enable/disable that plugin during testing.

Good point @vedanshujain. I've add testing instructions for situations where git and unit tests might not be available. Please take another look.

**Comment by lsinger** [comment]
> @lsinger I'm running the test as follows: `vendor/bin/phpunit tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php` (I always run the unit test locally like that). Shouldn't this be equivalent to what you posted in the testing instructions?

@Konamiman *Should* -- maybe. ☺️ Unfortunately our unit tests aren't well isolated from each other, so there's definitely a chance that just running the single test file is an issue. We can't rule out anything being off in your local setup, though -- that's what I like about the ephemerality of wp-env. FWIW, the pnpm script runs this command in the container: `vendor/bin/phpunit -c phpunit.xml --verbose`

Maybe adding the configuration file argument explicitly makes a difference?

File: plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php [hunk 1]
```diff
@@ -3260,4 +3260,49 @@ public function test_ephemeral_meta_updates() {
...
+
+		$order = WC_Helper_Order::create_order();
+		$order->save();
+		$order_id = $order->get_id();
+
+		$wpdb->query( "INSERT INTO {$order_meta_table} (order_id, meta_key, meta_value) VALUES ({$order_id}, '{$meta_key}', '{$meta_value}')" ); // phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared,WordPress.DB.SlowDBQuery.slow_db_query_meta_key,WordPress.DB.SlowDBQuery.slow_db_query_meta_value
+		$order->set_date_modified( time() );
+		$order->save();
+
+		$order_results = wc_get_orders(
```
Comment (vedanshujain) [inline plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php]: why not just `wc_get_order( $order->get_id() )`, that would make it more obvious to whats going on?

File: plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php [hunk 2]
```diff
@@ -3260,4 +3260,49 @@ public function test_ephemeral_meta_updates() {
...
+		$this->toggle_cot_authoritative( true );
+		$this->enable_cot_sync();
+		$order_meta_table = $this->sut::get_meta_table_name();
+
+		$order = WC_Helper_Order::create_order();
+		$order->save();
+		$order_id = $order->get_id();
+
+		$wpdb->query( "INSERT INTO {$order_meta_table} (order_id, meta_key, meta_value) VALUES ({$order_id}, '{$meta_key}', '{$meta_value}')" ); // phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared,WordPress.DB.SlowDBQuery.slow_db_query_meta_key,WordPress.DB.SlowDBQuery.slow_db_query_meta_value
+		$order->set_date_modified( time() );
```
Comment (vedanshujain) [inline plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php]: probably change this to `time() + 1` otherwise this may get flaky

**Review by vedanshujain — COMMENTED** [review]
Looks good, can you add one more test which checks for deleting meta as well, just like we are covering adding meta.

**Comment by Konamiman** [comment]
@lsinger What fixes the test for me is adding the following to the test:

```php
require_once __DIR__ . '/../../../../../legacy/unit-tests/webhooks/functions.php';
```

I think this makes sense, as we are referencing a file in `tests/legacy` from withing a file in `tests/php` - I would expect an explicit include to be required in this case.

**Comment by jorgeatorres** [comment]
@lsinger: While looking at this it occurred to me that instead of trying to capture the throwable and determine its nature by the error message, we could preemptively detect when `$metadata->value` is of type `__PHP_Incomplete_Class` and resort to the direct queries in that case?

I know it's not exactly the same logic but given we only seem to care about that one throwable, the result is pretty much the same (and maybe even bit cleaner).

There might be some limitation I'm ignoring, so please let me know if that's the case 😸.

**Comment by lsinger** [comment]
> @lsinger What fixes the test for me is adding the following to the test:
> 
> ```
> require_once __DIR__ . '/../../../../../legacy/unit-tests/webhooks/functions.php';
> ```
> 
> I think this makes sense, as we are referencing a file in `tests/legacy` from withing a file in `tests/php` - I would expect an explicit include to be required in this case.

@Konamiman interesting -- this should be a problem on `trunk` as well then? 🤔 Should we file an issue?

File: plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php (hunk 1 above)
Comment (lsinger) [reply plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php]: Good point -- this query was the one provided by the issue reporter, and I hadn't simplified it when it had become clear what the underlying issue is. Fixed. 👍

**Comment by lsinger** [comment]
> @lsinger: While looking at this it occurred to me that instead of trying to capture the throwable and determine its nature by the error message, we could preemptively detect when `$metadata->value` is of type `__PHP_Incomplete_Class` and resort to the direct queries in that case?
> 
> I know it's not exactly the same logic but given we only seem to care about that one throwable, the result is pretty much the same (and maybe even bit cleaner).
> 
> There might be some limitation I'm ignoring, so please let me know if that's the case 😸.

Thank you @jorgeatorres, that's a great suggestion -- I implemented it in 1abb8ed. There might of course be other cases where an exception is thrown, but we haven't come across them yet -- and the old solution would have rethrown them anyway. So I agree that we should prefer this simpler way of handling things.

**Comment by lsinger** [comment]
Addresses all review comments, would love to get another review. ☺️

File: plugins/woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php [hunk 3]
```diff
@@ -705,8 +705,11 @@ public function update_post_modified_data( $data, $postarr ) {
 	 * Helper method to update order metadata from initialized order object.
 	 *
 	 * @param WC_Abstract_Order $order Order object.
+	 * @throws \Throwable If adding post meta fails with both `add_post_meta` and `$wpdb->insert`.
```
Comment (jorgeatorres) [inline plugins/woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php]: We're no longer throwing an exception after the latest change.

File: plugins/woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php [hunk 4]
```diff
@@ -734,7 +737,25 @@ protected function update_order_meta_from_object( $order ) {
 					}
 				}
 			}
-			add_post_meta( $order->get_id(), $meta_data->key, $meta_data->value, false );
+			if ( 'object' === gettype( $meta_data->value ) && '__PHP_Incomplete_Class' === get_class( $meta_data->value ) ) {
```
Comment (jorgeatorres) [inline plugins/woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php]: I don't see anything wrong here, but maybe `is_object()` instead of `gettype()` would be a bit shorter?

File: plugins/woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php [hunk 5]
```diff
@@ -734,7 +737,25 @@ protected function update_order_meta_from_object( $order ) {
...
+						'post_id'    => $order->get_id(),
+						'meta_key'   => $meta_data->key, // phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_key
+						'meta_value' => $meta_value, // phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_value
+					),
+					array( '%d', '%s', '%s' )
+				);
+
+				if ( ! $result ) {
+					throw $th;
+				}
```
Comment (jorgeatorres) [inline plugins/woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php]: We no longer have anything to throw :)

File: plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php [hunk 6]
```diff
@@ -2880,8 +2880,11 @@ public function read_meta( &$object ) {
 	 * @param \stdClass $meta (containing at least ->id).
 	 *
 	 * @return bool
+	 * @throws \Throwable If deleting post meta fails with both `delete_post_meta` and `$wpdb->delete`.
```
Comment (jorgeatorres) [inline plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php]: Like above, this no longer throws an exception.

**Review by jorgeatorres — COMMENTED** [review]
@lsinger: Thanks for making the changes. I haven't properly tested this yet, but I noticed a few things in the code, and so I'm leaving some feedback. Luckily, nothing serious. Let me know what you think.

**Comment by lsinger** [comment]
> @lsinger: Thanks for making the changes. I haven't properly tested this yet, but I noticed a few things in the code, and so I'm leaving some feedback. Luckily, nothing serious. Let me know what you think.

Thank you for having my back, @jorgeatorres -- I should have caught these. All great notes / suggestions, I made the changes. ✅

**Comment by vedanshujain** [comment]
@jorgeatorres  @lsinger , re: the `__PHP_Incomplete_Class` change, I agree that it's way cleaner than matching the error message, but note that it's not functionally the same. Now we will trigger this `if` condition in earlier PHP versions where this was not an issue, but also, if `map_deep` ever changes to actually handle this scenario, we will still be hitting the `if` code block even if it's not necessary anymore. 

It's still should be fine, unless you can think of a scenario when it isn't.

**Comment by lsinger** [comment]
> @jorgeatorres @lsinger , re: the `__PHP_Incomplete_Class` change, I agree that it's way cleaner than matching the error message, but note that it's not functionally the same. Now we will trigger this `if` condition in earlier PHP versions where this was not an issue, but also, if `map_deep` ever changes to actually handle this scenario, we will still be hitting the `if` code block even if it's not necessary anymore.
> 
> It's still should be fine, unless you can think of a scenario when it isn't.

I can't think of a scenario where it wouldn't be fine to go the alternative route. Note also that we do log the instances where we go that way, so it wouldn't be able to happen completely invisibly.

**Comment by jorgeatorres** [comment]
Hey @vedanshujain!

> but note that it's not functionally the same.

I agree, that's why I said that it was mostly but not exactly "the same".

> Now we will trigger this `if` condition in earlier PHP versions where this was not an issue but also, if `map_deep` ever changes to actually handle this scenario.

We could mitigate this bit by adding a PHP version check, but not sure if it's worth it. As for `map_deep()` eventually handling this, that'd be great but I don't think that will happen very soon and (even in that case) we would need to have a work around in place.
I'm not sure if error messages are "guaranteed" to remain the same either, so relying on that seems (to me) even less future proof.

> It's still should be fine, unless you can think of a scenario when it isn't.

I can't think of one right now but "never say never". Considering this is already an edge case, I'd expect this new code not be hit very often so we would at least have some room to improve it if necessary.

---

> Note also that we do log the instances where we go that way, so it wouldn't be able to happen completely invisibly.

@lsinger: Could you expand on this? I don't think we're logging anything now that we would handle the exception before it even occurs. Maybe actually logging this unusual instance of meta sync wouldn't be a bad idea.

**Comment by lsinger** [comment]
> @lsinger: Could you expand on this? I don't think we're logging anything now that we would handle the exception before it even occurs. Maybe actually logging this unusual instance of meta sync wouldn't be a bad idea.

🤦 No, we're not -- I was thinking of #43739. So let me rephrase: we should log this. 💯 I'll add that.

**Comment by vedanshujain** [comment]
Ok sounds good, let's ship then!

**Comment by lsinger** [comment]
@jorgeatorres @vedanshujain added logging in [0beb5d5](https://github.com/woocommerce/woocommerce/pull/43517/commits/0beb5d5db17d20a10bea2b7bdac4f5ee3b67356a). 

> Ok sounds good, let's ship then!

FWIW, someone needs to actually hit the "approve" button. 😅 

Screenshot 2024-01-25 at 12 30 54

**Comment by Konamiman** [comment]
> this should be a problem on trunk as well then? 🤔 Should we file an issue?

I rather think that when running the full test suite (or with `pnpm` as you suggest) the missing file gets included somehow (via another test?) and the issue only shows when running the tests for `OrdersTableDataStoreTest.php` only. So I'd add that `require_once` statement as it won't hurt anyway.

**Comment by lsinger** [comment]
> > this should be a problem on trunk as well then? 🤔 Should we file an issue?
> 
> I rather think that when running the full test suite (or with `pnpm` as you suggest) the missing file gets included somehow (via another test?) and the issue only shows when running the tests for `OrdersTableDataStoreTest.php` only. So I'd add that `require_once` statement as it won't hurt anyway.

@Konamiman my intuition is that we should create an individual PR for this as it's fixing a different issue. What do you think?

**Review by vedanshujain — APPROVED** [review]
